### DISCUSSION TITLE
Dakota source needs changes for numpy 2.0

### DISCUSSION THREAD

**Discussion by wvangeit** [snl-dakota/dakota#131] [Ideas]
Numpy 2.0 has removed some deprecated API. Compiling dakota with numpy enabled is broken atm (6.19.0 and 6.20.0).
It's throwing errors like:
```
/project/dakota/src/PythonInterface.cpp: In member function ‘bool Dakota::PythonInterface::python_convert_int(const ArrayT&, Size, PyObject**)’:
    /project/dakota/src/PythonInterface.cpp:334:45: error: ‘PyArray_INT’ was not declared in this scope; did you mean ‘PyArray_Put’?
      334 |     if (!(*dst = PyArray_SimpleNew(1, dims, PyArray_INT))) {
          |                                             ^~~~~~~~~~~
    /tmp/pip-build-env-7kihbjzh/overlay/lib/python3.11/site-packages/numpy/_core/include/numpy/ndarrayobject.h:119:46: note: in definition of macro ‘PyArray_SimpleNew’
      119 |         PyArray_New(&PyArray_Type, nd, dims, typenum, NULL, NULL, 0, 0, NULL)
          |                                              ^~~~~~~
    /project/dakota/src/PythonInterface.cpp: In member function ‘bool Dakota::PythonInterface::python_convert(const Dakota::RealVector&, PyObject**)’:
    /project/dakota/src/PythonInterface.cpp:365:45: error: ‘PyArray_DOUBLE’ was not declared in this scope; did you mean ‘PyArray_DTYPE’?
      365 |     if (!(*dst = PyArray_SimpleNew(1, dims, PyArray_DOUBLE))) {
          |                                             ^~~~~~~~~~~~~~
```

Part of it can be solved with this script: https://github.com/numpy/numpy/blob/main/tools/replace_old_macros.sed
but not everything.

**Comment by jadamstephens** [comment]
Thanks for this report. I made a commit last week that does a few things.

* Upgrades our Pybind11 a bit to support NumPy 2
* Default disables our "legacy" Python inferface when building from source, which used Python's and NumPy's C APIs directly. This interface is not compatible with NumPy 2, and I don't intend to fix it. It'll probably be removed soon because I think they only thing it provides that the Pybind11 version doesn't is Python 2 support.
* Added a CMake variable to provide more granular control over whether the legacy interface is enabled.

The changes are available now in the repo and of course will be part of the 6.21, which will come out in 2.5 weeks or so.

**Comment by wvangeit** [comment]
Thank you @jadamstephens
